paasta_tools/autoscaling_lib.py

```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
-    # killing instance using the noTermination
-    # ExcessCapacityTerminationPolicy
-    #########
+def kill_instances_gracefully(instances_to_kill):
+    """Kills instances after waiting for slave to be running 0 mesos tasks
+
+    :param instances_to_kill: dict of instances to kill {<slave_id>: <instance_id>}"""
+    ec2_client = boto3.client('ec2')
+    try:
+        with Timeout(seconds=CLUSTER_DRAIN_TIMEOUT):
+            while instances_to_kill:
+                mesos_state = get_mesos_state_from_leader()
+                slaves_count = get_mesos_task_count_by_slave(mesos_state, pool=None)
+                for slave_id, instance_id in instances_to_kill.items():
+                    num_tasks = slaves_count[slave_id]['count']
+                    log.debug("Instance: {0}, still running {1} tasks".format(instance_id, num_tasks))
+                    if num_tasks == 0:
+                        ec2_client.termintate_instances(InstanceIds=[instance_id])
+                        instances_to_kill.pop(slave_id)
+                time.sleep(5)
+    except TimeoutError:
+        log.error("Timed out after {0} waiting to drain {1}, now terminating anyway".format(CLUSTER_DRAIN_TIMEOUT,
+                                                                                            instances_to_kill))
+        ec2_client.termintate_instances(InstanceIds=instances_to_kill)
+
+
+@register_autoscaling_component('aws_spot_fleet_request', SCALER_KEY)
+def scale_aws_spot_fleet_request(resource, delta, target_capacity, slaves_to_kill):
+    """Scales a spot fleet request by delta to reach target capacity
+    If scaling up we just set target capacity and let AWS take care of the rest
+    If scaling down we pick the slaves we'd prefer to kill, put them in maintenance
+    mode and drain them (via paasta_maintenance and setup_marathon_jobs). We then kill
+    them once they are running 0 tasks or once a timeout is reached
+
+    :param resource: resource to scale
+    :param delta: integer change in number of servers
+    :param target_capacity: target number of instances
+    :param slaves_to_kill: list of slave pids to kill"""
+    sfr_id = resource['id']
+    ec2_client = boto3.client('ec2')
+    if delta == 0:
+        return
+    elif delta > 0:
+        return ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
+                                                    ExcessCapacityTerminationPolicy='noTermination')
+    elif delta < 0:
+        slaves_by_ip = {slave: slave_pid_to_ip(slave) for slave in slaves_to_kill}
+        instances_to_kill = {slave_id: get_instance_id_from_ip(ip) for slave_id, ip in slaves_by_ip.items()}
+        start = time.time()
+        # This ensures we are in maintenance mode longer than the time before
+        # we give up trying to drain and just terminate
+        duration = CLUSTER_DRAIN_TIMEOUT + 300
+        drain(slaves_by_ip.values(), start, duration)
```

Comment:
please modify this function to return the http response and let the caller decide what to do, rather than relying on it's side-effecty behaviour